Index: jdk8u51-b16/jdk/src/share/native/sun/awt/splashscreen/splashscreen_gif.c
===================================================================
--- jdk8u51-b16.orig/jdk/src/share/native/sun/awt/splashscreen/splashscreen_gif.c
+++ jdk8u51-b16/jdk/src/share/native/sun/awt/splashscreen/splashscreen_gif.c
@@ -309,8 +309,14 @@ SplashDecodeGif(Splash * splash, GifFile
 
     free(pBitmapBits);
     free(pOldBitmapBits);
-
+#if GIFLIB_MAJOR > 5 || (GIFLIB_MAJOR == 5 && GIFLIB_MINOR >= 1)
+    int error = 0;
+    DGifCloseFile(gif, &error);
+    if (error)
+        return 0;
+#else
     DGifCloseFile(gif);
+#endif
 
     return 1;
 }
@@ -318,9 +324,16 @@ SplashDecodeGif(Splash * splash, GifFile
 int
 SplashDecodeGifStream(Splash * splash, SplashStream * stream)
 {
+#ifdef GIFLIB_MAJOR >= 5
+    int error = 0;
+    GifFileType *gif = DGifOpen((void *) stream, SplashStreamGifInputFunc, &error);
+    if (error)
+        return 0;
+#else
     GifFileType *gif = DGifOpen((void *) stream, SplashStreamGifInputFunc);
 
     if (!gif)
         return 0;
+#endif
     return SplashDecodeGif(splash, gif);
 }
