Index: jdk8u152-b16/common/autoconf/generated-configure.sh
===================================================================
--- jdk8u152-b16.orig/common/autoconf/generated-configure.sh
+++ jdk8u152-b16/common/autoconf/generated-configure.sh
@@ -29731,7 +29731,7 @@ fi
   #
   case $COMPILER_NAME in
     gcc )
-      COMMON_CCXXFLAGS_JDK="$COMMON_CCXXFLAGS $COMMON_CCXXFLAGS_JDK -W -Wall -Wno-unused -Wno-parentheses \
+      COMMON_CCXXFLAGS_JDK="$COMMON_CCXXFLAGS $COMMON_CCXXFLAGS_JDK -W -Wno-sign-compare -Wno-deprecated-declarations \
       -pipe \
       -D_GNU_SOURCE -D_REENTRANT -D_LARGEFILE64_SOURCE"
       CXXSTD_CXXFLAG="-std=gnu++98"
Index: jdk8u152-b16/common/autoconf/toolchain.m4
===================================================================
--- jdk8u152-b16.orig/common/autoconf/toolchain.m4
+++ jdk8u152-b16/common/autoconf/toolchain.m4
@@ -1018,7 +1018,7 @@ AC_DEFUN_ONCE([TOOLCHAIN_SETUP_COMPILER_
   #
   case $COMPILER_NAME in
     gcc )
-      COMMON_CCXXFLAGS_JDK="$COMMON_CCXXFLAGS $COMMON_CCXXFLAGS_JDK -W -Wall -Wno-unused -Wno-parentheses \
+      COMMON_CCXXFLAGS_JDK="$COMMON_CCXXFLAGS $COMMON_CCXXFLAGS_JDK -W -Wno-sign-compare -Wno-deprecated-declarations \
       -pipe \
       -D_GNU_SOURCE -D_REENTRANT -D_LARGEFILE64_SOURCE"
       CXXSTD_CXXFLAG="-std=gnu++98"
