Index: jdk8u-tuxjdk/hotspot/make/linux/makefiles/adjust-mflags.sh
===================================================================
--- jdk8u-tuxjdk.orig/hotspot/make/linux/makefiles/adjust-mflags.sh
+++ jdk8u-tuxjdk/hotspot/make/linux/makefiles/adjust-mflags.sh
@@ -64,7 +64,7 @@ MFLAGS=`
 	echo "$MFLAGS" \
 	| sed '
 		s/^-/ -/
-		s/ -\([^ 	][^ 	]*\)j/ -\1 -j/
+		s/ -\([^ 	I][^ 	I]*\)j/ -\1 -j/
 		s/ -j[0-9][0-9]*/ -j/
 		s/ -j\([^ 	]\)/ -j -\1/
 		s/ -j/ -j'${HOTSPOT_BUILD_JOBS:-${default_build_jobs}}'/
