Index: jdk8u152-b16/hotspot/src/share/vm/runtime/arguments.cpp
===================================================================
--- jdk8u152-b16.orig/hotspot/src/share/vm/runtime/arguments.cpp
+++ jdk8u152-b16/hotspot/src/share/vm/runtime/arguments.cpp
@@ -1572,11 +1572,7 @@ void Arguments::set_conservative_max_hea
 
 void Arguments::select_gc_ergonomically() {
   if (os::is_server_class_machine()) {
-    if (should_auto_select_low_pause_collector()) {
       FLAG_SET_ERGO(bool, UseConcMarkSweepGC, true);
-    } else {
-      FLAG_SET_ERGO(bool, UseParallelGC, true);
-    }
   }
 }
 
